### PR TITLE
Encapsulate WP functions calls [MAILPOET-1826]

### PR DESCRIPTION
**Pull request by amine-mp**
[MAILPOET-1826]

[MAILPOET-1826]: https://mailpoet.atlassian.net/browse/MAILPOET-1826

### CODE REVIEW

File: lib/Newsletter/Editor/PostTransformer.php [hunk 1]
```diff
@@ -138,6 +139,25 @@ private function getContent($post, $with_post_class, $display_type) {
...
+    global $content_width; // default is NULL
+    
+    $content_width_copy = $content_width;
+    $content_width = Env::NEWSLETTER_CONTENT_WIDTH;
+    $image_info = $this->wp->wpGetAttachmentImageSrc($id, 'mailpoet_newsletter_max');
+    $content_width = $content_width_copy;
+
+    return $image_info;
+  }
+
```
Comment (JanJakes) [inline lib/Newsletter/Editor/PostTransformer.php]: The commit message says `getImageInfo` was moved but I don't see it removed anywhere. What's the context/reason for this change?

File: lib/WP/Functions.php [hunk 2]
```diff
@@ -84,26 +95,379 @@ function wpEncodeEmoji() {
...
+
+  function isRtl() {
+    return call_user_func_array('is_rtl', func_get_args());
+  }
+
   function currentUserCan() {
     return call_user_func_array('current_user_can', func_get_args());
   }
 
-  function getImageInfo($id) {
```
Comment (JanJakes) [inline lib/WP/Functions.php]: Ok, now I see where it was removed from. Would be clearer if the move was in a single commit indeed. No need to change it now just noting we should pay more attention to that.

Comment (JanJakes) [reply lib/WP/Functions.php]: And I guess I also understand the move now - it's not a WP function wrapper so it shouldn't have been here, right?

File: lib/WP/Functions.php [hunk 3]
```diff
@@ -1,9 +1,20 @@
 <?php
 namespace MailPoet\WP;
 
-use MailPoet\Config\Env;
-
 class Functions {
+  static private $instance;
+
+  static function get() {
+    if (self::$instance == null) {
```
Comment (JanJakes) [inline lib/WP/Functions.php]: `==` => `===`

File: lib/WP/Functions.php [hunk 4]
```diff
@@ -84,26 +95,379 @@ function wpEncodeEmoji() {
     return call_user_func_array('wp_encode_emoji', func_get_args());
   }
 
+  function __() {
+    return call_user_func_array('__', func_get_args());
+  }
+
+  function stripslashesDeep() {
+    return call_user_func_array('stripslashes_deep', func_get_args());
+  }
```
Comment (JanJakes) [inline lib/WP/Functions.php]: So I see this class for the first time and it seems we're using `call_user_func_array` everywhere. I understand it's easier to do for the initial refactoring stage but I also think we really should annotate the methods properly & pass parameters by variables. It will be much more convenient for use in IDEs, for devs when looking into the code, and for PHPStan. Should we create a ticket for that?

File: lib/WP/Functions.php [hunk 5]
```diff
@@ -470,4 +470,16 @@ function getPostTypes() {
...
+    return call_user_func_array('get_editable_roles', func_get_args());
+  }
+
+  function getCategories() {
+    return call_user_func_array('get_categories', func_get_args());
+  }
+
+  function hasFilter() {
+    return call_user_func_array('has_filter', func_get_args());
+  }
```
Comment (JanJakes) [inline lib/WP/Functions.php]: This is a huge commit that I'm unable to review completely and I believe you've use some automatization to replace all of this. That's why it would be nice next time if the commit was really atomic and did only what it says in the commit message. Here is some other code appended but if there is any other code in other places I'm unable to discover that.

File: lib/Models/ScheduledTask.php [hunk 6]
```diff
@@ -1,7 +1,6 @@
 <?php
 
 namespace MailPoet\Models;
-
```
Comment (JanJakes) [inline lib/Models/ScheduledTask.php]: This commit has also many unrelated whitespace coding style changes. Why? They shouldn't be part of this commit, probably not even of this pull request.

File: lib/Subscribers/ImportExport/PersonalDataExporters/NewsletterClicksExporter.php [hunk 7]
```diff
@@ -1,9 +1,9 @@
 <?php
-
```
Comment (JanJakes) [inline lib/Subscribers/ImportExport/PersonalDataExporters/NewsletterClicksExporter.php]: Another very common unrelated coding style change in this commit. I don't even think it's our coding style to have that empty line removed.

File: lib/Cron/DaemonHttpRunner.php [hunk 8]
```diff
@@ -1,8 +1,11 @@
 <?php
 namespace MailPoet\Cron;
 
+use MailPoet\WP\Functions as WPFunctions;
+
 if (!defined('ABSPATH')) exit;
 
+
```
Comment (JanJakes) [inline lib/Cron/DaemonHttpRunner.php]: Another very common unrelated coding style change.

File: lib/Config/AccessControl.php [hunk 9]
```diff
@@ -14,48 +14,41 @@ class AccessControl {
   const PERMISSION_MANAGE_SEGMENTS = 'mailpoet_manage_segments';
   const NO_ACCESS_RESTRICTION = 'mailpoet_no_access_restriction';
 
-  /** @var WPFunctions */
-  private $wp;
-
-  function __construct(WPFunctions $wp) {
-    $this->wp = $wp;
-  }
-
```
Comment (JanJakes) [inline lib/Config/AccessControl.php]: Why do you remove DI from a class where it was already used? I thought that the singleton "hack" is something temporary like a stage in between transforming all classes to DI. What do we "downgrade" from that here?

File: lib/Form/Block/Base.php [hunk 10]
```diff
@@ -4,6 +4,7 @@
 
 use MailPoet\Form\Util\FieldNameObfuscator;
 use MailPoet\Models\ModelValidator;
+use MailPoet\WP\Functions as WPFunctions;
```
Comment (JanJakes) [inline lib/Form/Block/Base.php]: What happened here? Was it broken before?

File: lib/WP/Functions.php [hunk 11]
```diff
@@ -482,4 +482,16 @@ function getCategories() {
...
+    return call_user_func_array('wc_price', func_get_args());
+  }
+
+  function wcGetOrder() {
+    return call_user_func_array('wc_get_order', func_get_args());
+  }
+  function wcGetCustomerOrderCount() {
+    return call_user_func_array('wc_get_customer_order_count', func_get_args());
+  }
+
```
Comment (JanJakes) [inline lib/WP/Functions.php]: How is this related to this commit?

File: tests/integration/WP/FunctionsTest.php [hunk 12]
```diff
@@ -42,25 +42,6 @@ function makeAttachment($upload, $parent_post_id = 0) {
...
-    $id = $this->makeAttachment($upload);
-    expect($id)->notEmpty();
-
-    $image = $this->wp->getImageInfo($id);
-    expect($image[1])->equals(Env::NEWSLETTER_CONTENT_WIDTH);
-
-    wp_delete_attachment($id, $force_delete = true);
-  }
-
-
```
Comment (JanJakes) [inline tests/integration/WP/FunctionsTest.php]: Why was this test removed here?

File: lib/Config/Updater.php [hunk 13]
```diff
@@ -28,7 +28,7 @@ function __construct($plugin_name, $slug, $version) {
   function init() {
     WPFunctions::get()->addFilter('pre_set_site_transient_update_plugins', array($this, 'checkForUpdate'));
   }
-
+
```
Comment (JanJakes) [inline lib/Config/Updater.php]: Unrelated typo.

File: lib/Newsletter/Renderer/Renderer.php [hunk 14]
```diff
@@ -21,7 +21,10 @@ class Renderer {
   private $template;
   const NEWSLETTER_TEMPLATE = 'Template.html';
   const FILTER_POST_PROCESS = 'mailpoet_rendering_post_process';
-
+
```
Comment (JanJakes) [inline lib/Newsletter/Renderer/Renderer.php]: Unwanted whitespaces.

File: lib/WP/Functions.php [hunk 15]
```diff
@@ -154,7 +158,10 @@ function isRtl() {
   function currentUserCan() {
     return call_user_func_array('current_user_can', func_get_args());
   }
-
+
```
Comment (JanJakes) [inline lib/WP/Functions.php]: Unwanted whitespaces.

File: lib/WP/Functions.php [hunk 16]
```diff
@@ -154,7 +158,10 @@ function isRtl() {
   function currentUserCan() {
     return call_user_func_array('current_user_can', func_get_args());
   }
-
+  
+  /**
+   * @return \WP_Role | null
```
Comment (JanJakes) [inline lib/WP/Functions.php]: In other places we don't have whitespaces around `|`.

File: lib/Newsletter/Editor/PostTransformer.php (hunk 1 above)
Comment (amine-mp) [reply lib/Newsletter/Editor/PostTransformer.php]: This method was part of `WP\Functions` and is only used on the `PostTransformer` class. Since I wanted the `WPFunctions` class to only contains WP functions, I moved it here.

File: lib/WP/Functions.php (hunk 2 above)
Comment (amine-mp) [reply lib/WP/Functions.php]: Exactly!

File: lib/WP/Functions.php (hunk 3 above)
Comment (amine-mp) [reply lib/WP/Functions.php]: Yes, I will edit that.

File: lib/WP/Functions.php (hunk 4 above)
Comment (amine-mp) [reply lib/WP/Functions.php]: I agree. This can be the next step.

File: lib/WP/Functions.php (hunk 5 above)
Comment (amine-mp) [reply lib/WP/Functions.php]: I was unable to do that manually so I did it via a script, but the script messed up some formatting (add/remove empty lines). And I was also unable to do it in multiple commits. I am sorry for that.

File: lib/Config/AccessControl.php (hunk 9 above)
Comment (amine-mp) [reply lib/Config/AccessControl.php]: I didn't think about it as a hack, it makes mocking methods during tests easier. Let's take the following example:

```php
class A {
  function methodA() {
    //...
    B::methodB();
  }
}

class B {
  static function methodB() {
    $c = new C;
    $c->methodC();
    wp_call();
  }
}

class C {
  function methodC() {
    wp_call(...);
  }
}
```

If we want to use DI, we will need to:

- Add `$wp` as property of class `C` and inject it in the constructor.
- Add `$wp` as optional argument to `B::methodB` and do `new C($wp)`. 
- Add `$wp` as property of class `A` and inject it in the constructor. Even if class `A` doesn't call any WP function. But if we don't do that, we will not be able to mock WP calls when testing `A::methodA`.

This is just a small example, if we want to do that in our code, we will add `$wp` to the majority of our classes!

File: lib/Form/Block/Base.php (hunk 10 above)
Comment (amine-mp) [reply lib/Form/Block/Base.php]: Yes, my script didn't handle the `__` function well.

File: lib/WP/Functions.php (hunk 11 above)
Comment (amine-mp) [reply lib/WP/Functions.php]: This fixes the tests of the premium plugin. But I agree it should have been added in a separate commit.

File: tests/integration/WP/FunctionsTest.php (hunk 12 above)
Comment (amine-mp) [reply tests/integration/WP/FunctionsTest.php]: because the `getImageInfo` is no longer in this class. it's tested implicitly when testing methods of the `PostTransformer` class.

File: lib/WP/Functions.php (hunk 16 above)
Comment (amine-mp) [reply lib/WP/Functions.php]: I will change it

File: lib/WP/Functions.php (hunk 5 above)
Comment (JanJakes) [reply lib/WP/Functions.php]: It's OK to have the result of the script in one single commit, I was just pointing out these methods above seem to be added manually so it seems to me kind of mixed-in.

File: lib/Config/AccessControl.php (hunk 9 above)
Comment (JanJakes) [reply lib/Config/AccessControl.php]: Well, you could say that about every single class managed by DI and then we would not need DI at all :) It's very easy to use static service locator and just use it everywhere. But it's also not very good. It's also very easy to have everything global like WP does and that's even less good.

Adding a setter for tests (or anything else) is bad, using it has global side effects. In fact adding anything just for tests is bad, it means it is not a well-designed testable unit.

In your example I see many problems from my point of view - class `A` you should not use `B::methodB()` to initialize `B`, either the service or a non-static factory should be injected, `B` should not have a static factory method for `C`, if it is a factory, then normal method would do it, so it's hard to imagine such situation.
